generator/internal/metadata_decorator_rest_generator.cc

```diff
@@ -0,0 +1,270 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "generator/internal/metadata_decorator_rest_generator.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "absl/memory/memory.h"
+#include "absl/strings/str_split.h"
+#include "generator/internal/codegen_utils.h"
+#include "generator/internal/predicate_utils.h"
+#include "generator/internal/printer.h"
+#include <google/protobuf/descriptor.h>
+#include <algorithm>
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+namespace {
+
+enum ContextType { kPointer, kReference };
+
+std::string SetMetadataText(google::protobuf::MethodDescriptor const& method,
+                            ContextType context_type) {
+  std::string const context =
+      context_type == kPointer ? "*rest_context" : "rest_context";
+
+  auto info = ParseExplicitRoutingHeader(method);
+  if (info.empty()) {
+    return absl::StrFormat("  SetMetadata(%s);", context);
+  }
+
+  // clang-format off
+  std::string text;
+  text += "  std::vector<std::string> params;\n";
+  text += "  params.reserve(" + std::to_string(info.size()) + ");\n\n";
+  for (auto const& kv : info) {
+    // In the simplest (and probably most common) cases where no regular
+    // expression matching is needed for a given routing parameter key, we skip
+    // the static loading of `RoutingMatcher`s and simply use if statements.
+    if (std::all_of(
+            kv.second.begin(), kv.second.end(),
+            [](RoutingParameter const& rp) { return rp.pattern == "(.*)"; })) {
+      auto const* sep = "  ";
+      for (auto const& rp : kv.second){
+        text += sep;
+        text += "if (!request." + rp.field_name + "().empty()) {\n";
+        text += "    params.push_back(\"" + kv.first + "=\" + request." + rp.field_name + "());\n";
+        text += "  }";
+        sep = " else ";
+      }
+      text += "\n\n";
+      continue;
+    }
+    text += "  static auto* " + kv.first + "_matcher = []{\n";
+    text += "    return new google::cloud::internal::RoutingMatcher<$request_type$>{\n";
+    text += "      \"" + kv.first + "=\", {\n";
+    for (auto const& rp : kv.second) {
+      text += "      {[]($request_type$ const& request) -> std::string const& {\n";
+      text += "        return request." + rp.field_name + "();\n";
+      text += "      },\n";
+      // In the special match-all case, we do not bother to set a regex.
+      if (rp.pattern == "(.*)") {
+        text += "      absl::nullopt},\n";
+      } else {
+        text += "      std::regex{\"" + rp.pattern + "\", std::regex::optimize}},\n";
+      }
+    }
+    text += "      }};\n";
+    text += "  }();\n";
+    text += "  " + kv.first + "_matcher->AppendParam(request, params);\n\n";
+  }
+  text += "  if (params.empty()) {\n";
```

Comment:
No action required: I notice that both the gRPC and REST functions generate this `if(params.empty()) else ()` code, maybe we should add a generated `SetMetadata(rest_internal::RestContext& context, std::vector<std::string> params)` helper, and call it from here.

Reply:
Eh, there is one for the joined param list.

I am wondering if we should factor out the making of `std::string text`. It is identical to the gRPC version.